fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexBuildState.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * IndexBuildState.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.IndexState;
+import com.apple.foundationdb.record.metadata.Index;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A class that contains the build progress of a given index in a given record store.
+ * <p>
+ * It first contains {@link IndexState}. If the state is {@link IndexState#WRITE_ONLY}, it will also contains the number
+ * of records scanned and the estimated total number of records that need to be scanned.
+ * </p>
+ * <p>
+ * Note that index build state (especially records scanned) serves a informational purpose only in terms that records scanned is not necessarily 100% of records in total when the index build completes in several reasons:
+ * </p>
+ * <ul>
+ *     <li> Records scanned does not include the records scanned in the {@link OnlineIndexer} whose {@link OnlineIndexer.Builder#setTrackProgress(boolean)} is set to false. </li>
+ *     <li> Records added or deleted during the index build progress are not included in records scanned but are in records in total. </li>
+ *     <li> Records in total uses the number of all records in the store. In some cases, the real number of records that need to be scanned is less than this.</li>
+ * </ul>
+ * @see #getIndexBuildStateAsync(FDBRecordStore, Index)
+ */
+@API(API.Status.UNSTABLE)
+public class IndexBuildState {
+    @Nonnull
+    private final IndexState indexState;
+    @Nullable
+    private final Long recordsScanned;
+    @Nullable
+    private final Long recordsInTotal;
+
+    /**
+     * Get the build progress ({@link IndexBuildState}) of the given index in the given record store asynchronously.
+     * @param store the record store containing the index
+     * @param index the index needed to be checked
+     * @return a future that completes to the index build state
+     */
+    public static CompletableFuture<IndexBuildState> getIndexBuildStateAsync(FDBRecordStore store, Index index) {
```

Comment:
I think `FDBRecordStore` (here and elsewhere) could be `FDBRecordStoreBase<?>` (to avoid us taking an opinion here on which implementation of the interface is used).